server/src/main/java/org/elasticsearch/search/fetch/subphase/InnerHitsPhase.java

```diff
@@ -84,7 +88,24 @@ private void hitExecute(Map<String, InnerHitsContext.InnerHitSubContext> innerHi
             if (results == null) {
                 hit.setInnerHits(results = new HashMap<>());
             }
-            innerHitsContext.queryResult().topDocs(topDoc, innerHitsContext.sort() == null ? null : innerHitsContext.sort().formats);
+
+            final DocValueFormat[] sortValueFormats = innerHitsContext.sort() == null ? null : innerHitsContext.sort().formats;
+            List<RescoreContext> rescore = innerHitsContext.rescore();
+            if (rescore.isEmpty() == false) {
+                assert sortValueFormats == null;
+
+                TopDocs justTopDocs = topDoc.topDocs;
+                if (justTopDocs.scoreDocs.length > 0) {
+                    for (RescoreContext ctx : rescore) {
+                        justTopDocs = ctx.rescorer().rescore(justTopDocs, innerHitsContext.searcher(), ctx);
+                        // TODO: Assert that top docs are sorted by score
```

Comment:
This should be quite expensive assertion.

Reply:
++, but it's done for top-level rescore (see [here](https://github.com/elastic/elasticsearch/blob/4380cd1bd5892e3dc33d03ea3a8eb8f9c24da881/server/src/main/java/org/elasticsearch/search/rescore/RescorePhase.java#L50)), so it follows that we should probably do it for inner hits rescore as well.